maxfw/utils/image_functions.py

```diff
@@ -0,0 +1,539 @@
+
+        elif pic.ndim == 2:
+            # if 2D image, add channel dimension (HWC)
+            pic = np.expand_dims(pic, 2)
+
+    elif isinstance(pic, (bytes, bytearray)):
+        try:
+            # verify that the object can be loaded into memory
+            pic = np.array(Image.open(io.BytesIO(pic)))
+        except Exception:
+            raise TypeError('The input bytes object is not suitable for the Pillow library. Check the input again.')
+
+    npimg = pic
+    if not isinstance(npimg, np.ndarray):
+        raise TypeError('Input pic must be a bytes object or NumPy ndarray, ' +
+                        'not {}'.format(type(npimg)))
+
+    if npimg.shape[2] == 1:
+        expected_mode = None
+        npimg = npimg[:, :, 0]
+        if npimg.dtype == np.uint8:
+            expected_mode = 'L'
+        elif npimg.dtype == np.int16:
+            expected_mode = 'I;16'
+        elif npimg.dtype == np.int32:
+            expected_mode = 'I'
+        elif npimg.dtype == np.float32:
+            expected_mode = 'F'
+        if mode is not None and mode != expected_mode:
+            raise ValueError("Incorrect mode ({}) supplied for input type {}. Should be {}"
+                             .format(mode, np.dtype, expected_mode))
+        mode = expected_mode
+
+    elif npimg.shape[2] == 2:
+        permitted_2_channel_modes = ['LA']
+        if mode is not None and mode not in permitted_2_channel_modes:
+            raise ValueError("Only modes {} are supported for 2D inputs".format(permitted_2_channel_modes))
+
+        if mode is None and npimg.dtype == np.uint8:
+            mode = 'LA'
+
+    elif npimg.shape[2] == 4:
+        permitted_4_channel_modes = ['RGBA', 'CMYK', 'RGBX']
+        if mode is not None and mode not in permitted_4_channel_modes:
+            raise ValueError("Only modes {} are supported for 4D inputs".format(permitted_4_channel_modes))
+
+        if mode is None and npimg.dtype == np.uint8:
+            mode = 'RGBA'
+    else:
+        permitted_3_channel_modes = ['RGB', 'YCbCr', 'HSV']
+        if mode is not None and mode not in permitted_3_channel_modes:
+            raise ValueError("Only modes {} are supported for 3D inputs".format(permitted_3_channel_modes))
+        if mode is None and npimg.dtype == np.uint8:
+            mode = 'RGB'
+
+    if mode is None:
+        raise TypeError('Input type {} is not supported'.format(npimg.dtype))
+
+    # Verify that the target mode exists
+    assert target_mode in [1, 'L', 'P', 'RGB', 'RGBA', 'CMYK', 'YCbCr', 'LAB', 'HSV', 'I', 'F', 'RGBX', 'RGBBa']
+
+    return Image.fromarray(npimg, mode=mode).convert(target_mode)
+
+
+def pil_to_array(pic):
+    assert _is_pil_image(pic), 'The input image for `PILtoarray` is not a PIL Image object.'
+    return np.array(pic)
+
+
+def normalize(img):
+    if type(img) is not np.ndarray:
+        img = np.array(img)
+    return img / (np.max(img) - np.min(img))
+
+
+def standardize(img, mean=None, std=None):
```

Comment:
Looking at this (and the related tests) again, I don't think we can standardize a 4-channel image? E.g torchvision (https://github.com/pytorch/vision/blob/master/torchvision/transforms/functional.py#L191) only works on RGB (technically it would take 4 dims but is intended for RGB only).

Reply:
Yes, I don't think it's appropriate to standardize the `A` channel in the same way as the other channels. I will raise an error instead an update the test cases. Thank you!